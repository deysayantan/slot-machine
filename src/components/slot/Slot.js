import React, { Component } from 'react'
import { connect } from 'react-redux';
import Banana from '../../Assets/banana.jpg'
import Orange from '../../Assets/orange.jpg'
import Strawberry from '../../Assets/strawberry.jpg'
import Monkey from '../../Assets/monkey.jpg'
import { Styles } from '../styles/AppStyle';
import {saveResult} from '../../action'

const options = [1,2,3,4]
class Slot extends Component {
    constructor(props) {
        super(props)
        this.state = {
            label : this.sample(options),
            chosenImage : null
        }
    }
    componentDidMount(){
        this.chooseImage()
        this.startInterval(5000)
    }
    componentDidUpdate(prevProps,prevStates){
        if(prevProps.userInteraction !== this.props.userInteraction){
            clearInterval(this.interval)
            console.log(`${this.props.serial} ::: ${this.state.label}`)
            if(this.props.userInteraction){
                this.startInterval(50)
                setTimeout(()=> {
                    clearInterval(this.interval)
                    const payload = {
                        "key":this.props.serial,
                        "value":this.state.label
                    }
                    this.props.saveResult(payload)
                },10000) 
            }
            else{
                this.startInterval(5000)
                console.log("called")
            }
            
        }
    }
    componentWillUnmount(){
        clearInterval(this.interval)
    }
    startInterval = (msec) => {
        this.interval = setInterval(
            () => {
                this.setState({
                    label:this.sample(options)
                },()=>{
                    this.chooseImage()
                })
            }
        ,msec)
    }
    sample = arr => arr[Math.floor(Math.random() * arr.length)];
    chooseImage = () => {
        let image;
        switch(this.state.label){
            case 1:
                image = Banana
                break;
            case 2:
                image = Orange
                break;
            case 3:
                image = Monkey
                break;
            case 4:
                image = Strawberry
                break;
            default:
                image = null
                break;
        }
        this.setState({
            chosenImage: image
        })
    }
    render() {
        console.log(this.props)
        return (
            <img src={this.state.chosenImage} alt="icon" style={Styles.image} />
        )
    }
}

const mapStateToProps = ({ result }) => ({
    result,
});

const mapDispatchToProps = dispatch => ({
    saveResult: (payload) => dispatch(saveResult(payload)),
});

export default connect(
    mapStateToProps,
    mapDispatchToProps,
)(Slot);


