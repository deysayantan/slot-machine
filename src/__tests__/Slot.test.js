import React from 'react';
import ReactDOM from 'react-dom';
import Slot from '../components/slot/Slot';
import Adapter from 'enzyme-adapter-react-16'
import { configure,shallow } from 'enzyme'
import { Provider } from "react-redux";
import configureMockStore from "redux-mock-store";
const mockStore = configureMockStore();
const store = mockStore({});

configure({ adapter: new Adapter() })

const mockCallBack = jest.fn();
const slotElement = shallow((<Provider store={store}><Slot serial="slot1" userInteraction={mockCallBack} /></Provider>));

describe('Slot component snapshot',()=>{
    it('renders app module correctly',()=>{
        expect(slotElement).toMatchSnapshot();
    });
});
describe('Initial slot assignment',()=>{
    const componentInstance = slotElement.instance();
    const slot = new componentInstance.props.children.type.WrappedComponent(0)
    it('slot 1 value must be 1,2,3 or 4',()=>{
        expect([1, 2, 3, 4]).toContain(slot.state.label1)
    })
    it('slot 2 value must be 1,2,3 or 4',()=>{
        expect([1, 2, 3, 4]).toContain(slot.state.label2)
    })
    it('slot 3 value must be 1,2,3 or 4',()=>{
        expect([1, 2, 3, 4]).toContain(slot.state.label3)
    })
})
