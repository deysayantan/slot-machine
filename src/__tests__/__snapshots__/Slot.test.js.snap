// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slot component snapshot renders app module correctly 1`] = `ShallowWrapper {}`;
